server/extensions/handler.go

```diff
@@ -0,0 +1,93 @@
+package extensions
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/golang-jwt/jwt/v4"
+	log "github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type extension struct {
+	url     string
+	headers http.Header
+}
+
+type extensions map[string]extension
+
+type errorResponse struct {
+	Message string `json:"message"`
+}
+
+var httpClient = &http.Client{}
+
+const labelKey = "argocd.argoproj.io/extension"
+
+func NewHandler(ctx context.Context, kubernetesClient kubernetes.Interface, namespace string) (http.Handler, error) {
+
+	items, err := kubernetesClient.CoreV1().Secrets(namespace).List(ctx, metav1.ListOptions{LabelSelector: labelKey})
+	if err != nil {
+		return nil, err
+	}
+
+	extensions := extensions{}
+	for _, i := range items.Items {
+		name := i.Labels[labelKey]
+		log.WithField("name", name).Info("loading v2 extension")
+
+		e := extension{url: string(i.Data["url"]), headers: http.Header{}}
+
+		for k, v := range i.Data {
+			if strings.HasPrefix(k, "header.") {
+				e.headers.Add(strings.TrimPrefix(k, "header."), string(v))
+			}
+		}
+
+		extensions[name] = e
+	}
+
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+
+		ctx := r.Context()
+
+		name := strings.Split(strings.TrimPrefix(r.URL.Path, "/api/extensions/"), "/")[0]
```

Comment:
As far as I can tell so far, it's possible to have an extension with an empty string as the name. The extension would listen on `/api/extensions/`. That's fine I guess, but maybe we should mandate the name be a non-empty string so that we can use that endpoint later on if we want, for example, to serve a list of extensions.